src/ripple/conditions/rsa_sha256.h

```diff
@@ -0,0 +1,270 @@
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2016 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_CONDITIONS_RSA_SHA256_H
+#define RIPPLE_CONDITIONS_RSA_SHA256_H
+
+#include <ripple/conditions/condition.h>
+#include <ripple/conditions/fulfillment.h>
+#include <ripple/conditions/impl/base64.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/SecretKey.h>
+#include <ed25519-donna/ed25519.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/pem.h>
+#include <openssl/sha.h>
+#include <boost/algorithm/clamp.hpp>
+#include <boost/regex.hpp>
+#include <boost/optional.hpp>
+#include <algorithm>
+#include <iterator>
+
+namespace ripple {
+namespace cryptoconditions {
+
+struct rsa_sha256_t final
+    : public fulfillment_t
+{
+    struct rsa_deleter
+    {
+        void operator() (RSA* rsa) const
+        {
+            RSA_free (rsa);
+        }
+    };
+
+    using RsaKey = std::unique_ptr<RSA, rsa_deleter>;
+
+    Buffer modulus_;
+    Buffer signature_;
+
+    // Check whether the public modulus meets the requirements
+    // imposed by section 4.4.1 of the RFC.
+    static
+    bool
+    checkModulus (int len)
+    {
+        if (len <= 0)
+            return false;
+
+        return len == boost::algorithm::clamp(len, 128, 512);
+    }
+
+    bool sign_helper (RSA* key, Slice const& message)
+    {
+        int const keySize = RSA_size(key);
+        if (!checkModulus (keySize))
+            return false;
+
+        sha256_hasher h;
+        h (message.data(), message.size());
+        auto digest = static_cast<sha256_hasher::result_type>(h);
+
+        // Pad the result (-1 -> use hash length as salt length)
+        Buffer buf;
+
+        if (!RSA_padding_add_PKCS1_PSS(key,
+                buf.alloc(keySize), digest.data(),
+                EVP_sha256(), -1))
+            return false;
+
+        // Sign - we've manually padded the input already.
+        auto ret = RSA_private_encrypt(keySize, buf.data(),
+            signature_.alloc (buf.size()), key, RSA_NO_PADDING);
+
+        if (ret == -1)
+            return false;
```

Comment:
Clear `signature_` or we leave the object in an "invalid" state. (in quotes, but we should clean up our mess).

`signature_` is a `Buffer`. What do you think about making changes to local variables and let them "RIAA" away on error and only move the local vars into the member vars at the bottom of the function on success? We can use the same technique in the other functions where I complained about leaving the object in "messy" states.